File: 2023/UICPC DIV1/UICPC Round #03/solutions/G/papersnowflakes-zf.cpp
#include <bits/stdc++.h>

using namespace std;

using ll = long long;
using pll = pair<ll, ll>;
const ll INF = (1ll<<62);

int main() {
  ll n, m, l;
  cin >> n >> m >> l;

  multiset<pll> events;
  events.emplace(0, +1);
  for (ll i = 0, f; i < n; ++i) {
    cin >> f;
    ll loc = l+((i&1)?f:-f);
    events.emplace(loc, (i&1)?+2:-2);
    l = 2*loc-l;
  }
  events.emplace(l, (n&1)?+1:-1);
  for (ll i = 0, c; i < m; ++i) {
    cin >> c;
    events.emplace(c, 0);
  }
  events.emplace(INF, 0);

  ll pre = -INF, tot = 0, num = 0;
  for (auto p : events) {
    // cout << p.first << ' ' << p.second << endl;
    tot += (p.first - pre)*num;
    num += p.second;
    pre = p.first;
    if (p.second == 0) {
      cout << tot << (p.first == INF ? '\n' : ' ');
      tot = 0;
    }
  }

  return 0;
}

File: 2023/UICPC DIV1/UICPC Round #03/solutions/G/papersnowflakes-zf.py
#!/usr/bin/python3
n, m, l = map(int, input().split())

events = [(0,+1)]
i = 0
for f in map(int, input().split()):
    loc = l + (f if (i&1) else -f)
    events.append((loc, (+2 if (i&1) else -2)))
    l = 2*loc-l
    i += 1
events.append((l, (+1 if (n&1) else -1)))

events += [(int(c),0) for c in input().split()] + [(1<<62,0)]

events.sort()

pre, tot, num = 0, 0, 0
out = []
for x, d in events:
    pre, tot, num = x, tot + (x-pre)*num, num + d
    if not d:
        out.append(tot)
        tot = 0
print(*out)
